client/state/data-layer/wpcom-http/utils.js

```diff
@@ -110,54 +105,94 @@ const defaultOptions = {
  *
- * This function accepts three functions as the initiator, success,
- * and error handlers for actions and it will call the appropriate
- * one based on the stored meta. It accepts an optional fourth
- * function which will be called for progress events on upload.
+ * This function accepts several functions as the initiator, success, error and
+ * progress handlers for actions and it will call the appropriate
+ * one based on the stored meta.
+ *
+ * These handlers are action creators: based on the input data coming from the HTTP request,
+ * it will return an action (or an action thunk) to be executed as a response to the given
+ * HTTP event.
  *
  * If both error and response data is available this will call the
  * error handler in preference over the success handler, but the
  * response data will also still be available through the action meta.
  *
  * The functions should conform to the following type signatures:
- *   initiator  :: ReduxStore -> Action -> Dispatcher (middleware signature)
- *   onSuccess  :: ReduxStore -> Action -> Dispatcher -> ResponseData
- *   onError    :: ReduxStore -> Action -> Dispatcher -> ErrorData
- *   onProgress :: ReduxStore -> Action -> Dispatcher -> ProgressData
- *   fromApi    :: ResponseData -> [ Boolean, Data ]
+ *   initiator  :: Action -> Action (action creator with one Action parameter)
+ *   onSuccess  :: Action -> ResponseData -> Action (action creator with two params)
+ *   onError    :: Action -> ErrorData -> Action
+ *   onProgress :: Action -> ProgressData -> Action
+ *   fromApi    :: ResponseData -> TransformedData throws TransformerError|SchemaError
  *
  * @param {Function} initiator called if action lacks response meta; should create HTTP request
  * @param {Function} onSuccess called if the action meta includes response data
  * @param {Function} onError called if the action meta includes error data
- * @param {Object} options configures additional dispatching behaviors
- + @param {Function} [options.fromApi] maps between API data and Calypso data
- + @param {Function} [options.onProgress] called on progress events when uploading
- * @returns {?*} please ignore return values, they are undefined
+ * @param {Function} onProgress called on progress events when uploading
+ * @param {Function} fromApi maps between API data and Calypso data
+ * @returns {Action} action or action thunk to be executed in response to HTTP event
+ */
+export const dispatchRequest = options => {
+	if ( ! options.initiator ) {
+		warn( 'initiator handler is not defined: no request will ever be issued' );
+	}
+
+	if ( ! options.onSuccess ) {
+		warn( 'onSuccess handler is not defined: response to the request is being ignored' );
+	}
+
+	if ( ! options.onError ) {
+		warn( 'onError handler is not defined: error during the request is being ignored' );
+	}
+
+	return ( store, action ) => {
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
```

Comment:
is there a reason to split `createRequestAction` out into a separate function? It kind of seems like this function is doing nothing but validate the required parameters

Reply:
`createRequestAction` returns an optional action that can be also `undefined` -- in case of a `noop` handler. Then we check if there is anything to dispatch and call `dispatch` if it is. This logic would be harder to express inside one big function.

It's almost always a good thing if a big function can be split into two or more smaller ones.